File: backend/app/models/manufacturing.py
import uuid
import enum
from datetime import datetime, date
from decimal import Decimal

from sqlalchemy import (
    Column, String, Integer, Numeric, Boolean, Date, DateTime, Enum, ForeignKey, CheckConstraint
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.database import Base


class MOStatusEnum(enum.Enum):
    draft = "draft"
    confirmed = "confirmed"
    in_progress = "in_progress"
    to_close = "to_close"
    done = "done"
    cancelled = "cancelled"


class ManufacturingOrder(Base):
    __tablename__ = "manufacturing_orders"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    reference = Column(String(40), unique=True, nullable=False)
    finished_product_id = Column(
        UUID(as_uuid=True),
        ForeignKey("products.id"),
        nullable=False
    )
    bom_id = Column(
        UUID(as_uuid=True),
        ForeignKey("boms.id"),
        nullable=True
    )
    quantity = Column(Numeric(14, 3), nullable=False)
    assignee_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True
    )
    status = Column(
        Enum(MOStatusEnum, name="mo_status_enum"),
        nullable=False,
        default=MOStatusEnum.draft
    )
    auto_created = Column(Boolean, nullable=False, default=False)
    source_sales_order_id = Column(
        UUID(as_uuid=True),
        ForeignKey("sales_orders.id"),
        nullable=True
    )
    parent_mo_id = Column(
        UUID(as_uuid=True),
        ForeignKey("manufacturing_orders.id"),
        nullable=True
    )
    vin_number = Column(String(40), unique=True, nullable=True)
    scheduled_date = Column(Date, nullable=True)
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    created_by = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True
    )

    __table_args__ = (
        CheckConstraint("quantity > 0", name="manufacturing_orders_quantity_check"),
    )

    # Relationships
    finished_product = relationship(
        "Product",
        foreign_keys=[finished_product_id],
        back_populates="manufacturing_orders"
    )
    bom = relationship("BOM", back_populates="manufacturing_orders")
    assignee = relationship(
        "User",
        foreign_keys=[assignee_id],
        back_populates="manufacturing_orders_assigned"
    )
    created_by_user = relationship(
        "User",
        foreign_keys=[created_by],
        back_populates="manufacturing_orders_created"
    )
    source_sales_order = relationship(
        "SalesOrder",
        back_populates="manufacturing_orders_sourced"
    )
    parent_mo = relationship(
        "ManufacturingOrder",
        remote_side=[id],
        back_populates="child_mos"
    )
    child_mos = relationship(
        "ManufacturingOrder",
        back_populates="parent_mo"
    )
    components = relationship("MoComponent", back_populates="manufacturing_order", cascade="all, delete-orphan")
    work_orders = relationship("WorkOrder", back_populates="manufacturing_order", cascade="all, delete-orphan")


class MoComponent(Base):
    __tablename__ = "mo_components"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    mo_id = Column(
        UUID(as_uuid=True),
        ForeignKey("manufacturing_orders.id", ondelete="CASCADE"),
        nullable=False
    )
    component_product_id = Column(
        UUID(as_uuid=True),
        ForeignKey("products.id"),
        nullable=False
    )
    to_consume = Column(Numeric(14, 3), nullable=False)
    consumed_qty = Column(Numeric(14, 3), nullable=False, default=Decimal("0"))
    batch_number = Column(String(60), nullable=True)

    # Relationships
    manufacturing_order = relationship("ManufacturingOrder", back_populates="components")
    component_product = relationship("Product", back_populates="mo_components")


class WorkOrder(Base):
    __tablename__ = "work_orders"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    mo_id = Column(
        UUID(as_uuid=True),
        ForeignKey("manufacturing_orders.id", ondelete="CASCADE"),
        nullable=False
    )
    sequence = Column(Integer, nullable=False)
    operation_name = Column(String(120), nullable=False)
    work_center = Column(String(120), nullable=False)
    expected_duration_min = Column(Integer, nullable=False)
    real_duration_min = Column(Integer, nullable=True)
    pass_fail = Column(String(10), nullable=True)

    # Relationships
    manufacturing_order = relationship("ManufacturingOrder", back_populates="work_orders")

File: frontend/src/types/manufacturing.ts
export interface ProductBrief {
  id: string;
  name: string;
}

export interface UserBrief {
  id: string;
  name: string;
}

export interface MoComponent {
  id: string;
  component_product_id: string;
  component_product_name: string;
  to_consume: number;
  consumed_qty: number;
  batch_number?: string;
  free_to_use_qty?: number;
}

export interface WorkOrder {
  id: string;
  sequence: number;
  operation_name: string;
  work_center: string;
  expected_duration_min: number;
  real_duration_min?: number;
  pass_fail?: string;
}

export interface ManufacturingOrder {
  id: string;
  reference: string;
  finished_product: ProductBrief;
  bom_id?: string;
  quantity: number;
  status: 'draft' | 'confirmed' | 'in_progress' | 'done' | 'cancelled';
  auto_created: boolean;
  source_sales_order_id?: string;
  source_sales_order_ref?: string;
  assignee?: UserBrief;
  scheduled_date?: string;
  components: MoComponent[];
  work_orders: WorkOrder[];
  created_at: string;
  created_by?: UserBrief;
}

export interface ManufacturingOrderListItem {
  id: string;
  reference: string;
  finished_product_id: string;
  finished_product_name: string;
  quantity: number;
  status: string;
  auto_created: boolean;
  source_sales_order_id?: string;
  created_at: string;
}

export interface ManufacturingOrderCreateRequest {
  finished_product_id: string;
  quantity: number;
  bom_id?: string;
  assignee_id?: string;
  scheduled_date?: string;
}

export const MO_STATUS_COLORS: Record<string, string> = {
  draft: 'bg-slate-500/20 text-slate-300',
  confirmed: 'bg-blue-500/20 text-blue-300',
  in_progress: 'bg-violet-500/20 text-violet-300',
  done: 'bg-emerald-500/20 text-emerald-300',
  cancelled: 'bg-red-500/20 text-red-300',
};

export const MO_STATUS_LABELS: Record<string, string> = {
  draft: 'Draft',
  confirmed: 'Confirmed',
  in_progress: 'In Progress',
  done: 'Done',
  cancelled: 'Cancelled',
};
